processor/src/stack/overflow.rs

```diff
@@ -71,96 +75,132 @@ impl OverflowTable {
     // --------------------------------------------------------------------------------------------
 
     /// Pushes the specified value into the overflow table.
+    ///
+    /// Parameter clk specifies the clock cycle at which the value is added to the table.
     pub fn push(&mut self, value: Felt, clk: u64) {
-        // get the clock cycle of the row currently at the top of the overflow table. if the
-        // overflow table is empty, this is set to ZERO.
-        let prev = self
-            .active_rows
-            .last()
-            .map_or(ZERO, |&last_row_idx| self.all_rows[last_row_idx].clk);
+        // ZERO address indicates that the overflow table is empty, and thus, no actual value
+        // should be inserted into the table with this address. This is not a problem since for
+        // every real program, we first execution an operation marking the start of a code block,
+        // and thus, no operation can shift the stack to the right at clk = 0.
+        debug_assert_ne!(clk, 0, "cannot add value to overflow at clk=0");
 
         // create and record the new row, and also put it at the top of the overflow table
-        let row_idx = self.all_rows.len();
-        self.all_rows.push(OverflowTableRow::new(clk, value, prev));
-        self.active_rows.push(row_idx);
+        let row_idx = self.all_rows.len() as u32;
+        let new_row = OverflowTableRow::new(clk, value, self.last_row_addr);
+        self.all_rows.push(new_row);
+        self.active_rows.push(row_idx as usize);
+
+        // set the last row address to the address of the newly added row
+        self.last_row_addr = Felt::from(clk);
 
         // mark this clock cycle as the cycle at which a new row was inserted into the table
         self.update_trace
-            .push((clk, OverflowTableUpdate::RowInserted(row_idx as u32)));
+            .push((clk, OverflowTableUpdate::RowInserted(row_idx)));
 
         if self.trace_enabled {
             // insert a copy of the current table state into the trace
-            self.trace.insert(clk, self.get_values());
+            self.save_current_state(clk);
         }
     }
 
-    /// Removes the last value from the overflow table and returns it together with the clock
-    /// cycle of the next value in the table.
-    ///
-    /// If after the top value is removed the table is empty, the returned clock cycle is ZERO.
-    pub fn pop(&mut self, clk: u64) -> (Felt, Felt) {
+    /// Removes the last value from the overflow table and returns it.
+    pub fn pop(&mut self, clk: u64) -> Felt {
+        // if last_row_addr is ZERO, any rows in the overflow table are not accessible from the
+        // current context. Thus, we should not be able to remove them.
+        debug_assert_ne!(
+            self.last_row_addr, ZERO,
+            "overflow table is empty in the current context"
+        );
```

Comment:
This should actually never happen under normal execution because we call `Overflow::pop()` only from `Stack::shift_left()` function. And there, we remove an item from the overflow table only if the stack depth is > 16.